src/ucs/datastruct/usage_tracker.c

```diff
@@ -0,0 +1,295 @@
+
+    if (kh_resize(usage_tracker_hash, &usage_tracker->hash,
+                  active_capacity * 2) < 0) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_lru;
+    }
+
+    usage_tracker->ticks_per_flush = ticks_per_flush;
+    usage_tracker->active_capacity = active_capacity;
+    usage_tracker->active_thresh   = active_thresh;
+    usage_tracker->eject_thresh    = eject_thresh;
+    usage_tracker->flush_cb        = cb;
+    usage_tracker->opaque          = opaque;
+    usage_tracker->ticks           = 0;
+    *usage_tracker_p               = usage_tracker;
+    return UCS_OK;
+
+err_free_lru:
+    ucs_lru_destroy(usage_tracker->lru);
+err_free_tracker:
+    ucs_free(usage_tracker);
+err:
+    return status;
+}
+
+void ucs_usage_tracker_destroy(ucs_usage_tracker_h usage_tracker)
+{
+    ucs_lru_destroy(usage_tracker->lru);
+    kh_destroy_inplace(usage_tracker_hash, &usage_tracker->hash);
+    ucs_free(usage_tracker);
+}
+
+static ucs_usage_tracker_element_t *
+ucs_usage_tracker_put(ucs_usage_tracker_h usage_tracker, void *key)
+{
+    int ret;
+    khint_t iter;
+    ucs_usage_tracker_element_t *elem;
+
+    iter = kh_put(usage_tracker_hash, &usage_tracker->hash, (uint64_t)key,
+                  &ret);
+    ucs_assert(ret != UCS_KH_PUT_FAILED);
+
+    elem      = &kh_val(&usage_tracker->hash, iter);
+    elem->key = key;
+
+    if ((ret == UCS_KH_PUT_BUCKET_EMPTY) || (ret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        elem->hit_count = 0;
+        elem->tx        = 0;
+        elem->rx        = 0;
+        elem->active    = 0;
+    }
+
+    elem->hit_count++;
+    return elem;
+}
+
+ucs_status_t ucs_usage_tracker_get_score(ucs_usage_tracker_h usage_tracker,
+                                         void *key, size_t *score_p)
+{
+    ucs_usage_tracker_element_t *item;
+    khint_t iter;
+
+    iter = kh_get(usage_tracker_hash, &usage_tracker->hash, (uint64_t)key);
+    if (iter == kh_end(&usage_tracker->hash)) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    item     = &kh_value(&usage_tracker->hash, iter);
+    *score_p = ucs_usage_tracker_score(item);
+    return UCS_OK;
+}
+
+static ucs_usage_tracker_element_t *
+ucs_usage_tracker_get_min_active(ucs_usage_tracker_h usage_tracker)
+{
+    ucs_usage_tracker_element_t *min_item = NULL;
+    unsigned active_count                 = 0;
+    ucs_usage_tracker_element_t *item;
+    khint_t k;
+
+    for (k = kh_begin(&usage_tracker->hash); k != kh_end(&usage_tracker->hash);
+         ++k) {
+        if (!kh_exist(&usage_tracker->hash, k)) {
+            continue;
+        }
+
+        item = &kh_val(&usage_tracker->hash, k);
+
+        if (!item->active) {
+            continue;
+        }
+
+        active_count++;
+
+        if ((min_item == NULL) || (ucs_usage_tracker_score(item) <
+                                   ucs_usage_tracker_score(min_item))) {
+            min_item = item;
+        }
+    }
+
+    return (active_count == usage_tracker->active_capacity) ? min_item : NULL;

```

Comment:
I would remove the active count check outside this function.
we can define usage_tracker_is_full_capacity by checking kh_size(usage_tracker->hash)

Reply:
kh_size(usage_tracker->hash) does not count active entries. (it counts all entries).
I need to count only active.